Rewrite Ballot view

symroe opened the pull request:
This is a big clean up PR that replaces what is basically the most used and most complex view on the site.

The reasons for this are:

1. The view hadn't _really_ moved away from the old assumptions about `Post` objects. The ballot wasn't the main object that thing hung off, and lots of it needed a `Post` and `Election` object rather than a single `Ballot` object. Related, lots of logic was in helper functions when they could be on ballot manager methods.

2. The view used lots of helper methods that were useful when the data model was different, but aren't very "djangoy" and are complex to debug. As an example, the objects passed to templates were a combination of lists of dicts containing model properties rather than the models / querysets themselves.

3. The templates were hard to update or manage, and that matters as we're about the redesign them all.

symroe commented:
I reckon this is feature complete with the previous version now.

There are still a few things I want to improve, but I think this is a decent point to stop working, especially given that we're likely to change things as part of the redesign work (removing the inline candidate form, for example).

I think it's best to merge the Django 2.2 PR first, and I'll rebase over that.

chris48s commented:
I'll leave you to rebase this, get the tests passing and then I'll have a look at it.

In terms of what this is supposed to be doing, this is basically a straight refactor, yes? Goal is maintain the existing functionality, but clean things up.

I think the reason you originally got into this was trying to fix a problem related to https://trello.com/c/btANFyAH but its now not supposed to be addressing that issue at all, right?

chris48s commented:
I don't think its _exactly_ the same thing, but the failure you're seeing in CI was the same one I was getting in #997 before applying 68979d6 I wonder if this is a slightly different manifestation of the same underlying issue?

symroe commented:
Yeah, looks like 395efc3 fixes the underlying problem…

symroe commented:
> I think the reason you originally got into this was trying to fix a problem related to

Yep, that turned out to be to do with changing IDs over time, and the fact that the old ballot view [used the FK relationship between`Membership` and `Post`](https://github.com/DemocracyClub/yournextrepresentative/blob/68979d60f434dfdac39cc02d8dd4d41b44c5b704/ynr/apps/elections/views.py#L219).

Changing that relationship involved making quite a few changes, so I thought I'd bite the bullet and do the whole lot, even though the initial problem was fixed.

chris48s commented:
I've spent some time looking over it this afternoon, but there is a lot going on at once here.
I'm going to aim to post comments tomorrow.

chris48s commented on ynr/apps/elections/templates/elections/ballot_view.html:
```diff
@@ -31,92 +31,323 @@
+            </td>
+            <td>{{ candidate.party.name }}</td>
+            <td>
+              <button class="button tiny js-toggle-source-confirmation standing">Standing again</button>
```
A few lines up, we're only showing the "Actions" header if the user is authenticated, but then it looks like we're rendering the column/form content anyway, even if they're not.

chris48s commented on ynr/apps/elections/templates/elections/ballot_view.html:
```diff
@@ -31,92 +31,323 @@
+              </a>
+            </td>
+            <td>
+              <a class="button tiny js-toggle-source-confirmation standing">Actually, they are standing!</a>
```
Same issue here: A few lines up, we're only showing the "Actions" header if the user is authenticated, but then we're rendering the column/form content anyway, even if they're not.

chris48s commented on ynr/apps/elections/uk/templates/candidates/_candidates_for_post.html:
```diff
@@ -1,59 +1,6 @@
 {% load standing %}
 
```
Is the `_candidates_for_post.html` template actually used anywhere now? There's still a bunch of stuff in it, but I can't find anywhere its included. If not, can we delete it.

chris48s commented on ynr/apps/elections/templates/elections/ballot_view.html:
```diff
@@ -31,92 +31,323 @@
+            <p>Someone has suggested locking this ballot.</p>
+          {% endif %}
+        {% else %}
+          <form method=post id="suggest_lock_form" action="{% url 'constituency-suggest-lock' election_id=ballot.election %}">
```
```suggestion
          <form method="post" id="suggest_lock_form" action="{% url 'constituency-suggest-lock' election_id=ballot.election %}">
```

chris48s commented on ynr/apps/elections/templates/elections/ballot_view.html:
```diff
@@ -31,92 +31,323 @@
+                <img class="person-avatar" src="{{ candidate.person.get_display_image_url }}"/>
+                {% if candidate.result.is_winner or candidate.elected %}<strong>{% endif %}
+                  {{ candidate.person.name }}
+                {% if candidate.result.is_winner or candidate.elected %}</strong>{% endif %}
```
any reason not to write this block as:
```django
{% if candidate.result.is_winner or candidate.elected %}
  <strong>{{ candidate.person.name }}</strong>
{% endif %}
```
?

chris48s commented on ynr/apps/elections/templates/elections/ballot_view.html:
```diff
@@ -31,92 +31,323 @@
+    {% comment %}<h2>People not standing again</h2>
+
+    {{ people_not_standing }}
+{% endcomment %}
```
is this code in the `{% comment %}{% endcomment %}` supposed to be in or out? If its out, can we just delete it?

chris48s commented on ynr/apps/candidates/models/popolo_extra.py:
```diff
@@ -131,6 +155,72 @@ def has_results(self):
+            return True
+
+        # If a user is trusted to lock then they can edit memberships
+        # TODO: Is this right?
```
My inclination is that we should switch it so they have to unlock, but if you want to address it in another PR, cool.

symroe replied:
Yeah, #991 suggests changing it, but I reckon I'll do a PR after this

chris48s commented on ynr/apps/candidates/models/popolo_extra.py:
```diff
@@ -131,6 +155,72 @@ def has_results(self):
+        # https://github.com/DemocracyClub/yournextrepresentative/issues/991
+        if user.groups.filter(name=TRUSTED_TO_LOCK_GROUP_NAME).exists():
+            return True
+
```
Also, can we terminate this method with an explicit `return False` instead of relying on implicit `return None`

chris48s commented on ynr/apps/candidates/tests/test_candidacy_add_and_remove.py:
```diff
@@ -1,3 +1,5 @@
+from unittest import skip
```
unused import

chris48s commented on ynr/apps/candidates/tests/test_changes_for_review.py:
```diff
@@ -1,3 +1,5 @@
+from unittest import skip
```
unused import

chris48s commented on ynr/apps/elections/tests/data_timeline_helper.py:
```diff
@@ -0,0 +1,105 @@
+        self.assertDataTimelineNoResults(response)
+
+    def assertDataTimelineHasResults(self, response):
+        print(response.text)
```
This debug statement can be removed

chris48s commented on ynr/apps/elections/templates/elections/ballot_view.html:
```diff
@@ -31,92 +31,323 @@
+          </a>
+        {% endif %}
+
+      {% elif not request.user.is_authenticated %}
```
Any reason we're using `request.user` here but `user` elsewhere?

chris48s commented on ynr/apps/elections/views.py:
```diff
@@ -139,821 +125,51 @@ def dispatch(self, *args, **kwargs):
-                user=self.request.user, ballot=ballot
+                "current_user_suggested_lock"
+            ] = ballot.suggestedpostlock_set.filter(
+                user=self.request.user
```
This is throwing `'AnonymousUser' object is not iterable` if I'm not logged in. Can you reproduce?

symroe replied:
Hmm, no I can't. Can you give me more information about the set up?

chris48s replied:
Are you sure you're testing it with a ballot where `ballot.has_lock_suggestion == True`?

In fact, was it intentional to completely replace `self.request.user.is_authenticated` with `ballot.has_lock_suggestion` a few lines up?

here's a stack trace if that helps:

```
Traceback (most recent call last):
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/chris/repositories/github_chris48s/yournextrepresentative/ynr/apps/elections/helpers.py", line 35, in __call__
    return view(request, *args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/views/decorators/cache.py", line 31, in _cache_controlled
    response = viewfunc(request, *args, **kw)
  File "/home/chris/repositories/github_chris48s/yournextrepresentative/ynr/apps/elections/views.py", line 125, in dispatch
    return super().dispatch(*args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/views/generic/base.py", line 97, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/views/generic/base.py", line 158, in get
    context = self.get_context_data(**kwargs)
  File "/home/chris/repositories/github_chris48s/yournextrepresentative/ynr/apps/elections/views.py", line 153, in get_context_data
    user=self.request.user
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/manager.py", line 82, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/query.py", line 892, in filter
    return self._filter_or_exclude(False, *args, **kwargs)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/query.py", line 910, in _filter_or_exclude
    clone.query.add_q(Q(*args, **kwargs))
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1290, in add_q
    clause, _ = self._add_q(q_object, self.used_aliases)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1318, in _add_q
    split_subq=split_subq, simple_col=simple_col,
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1224, in build_filter
    self.check_related_objects(join_info.final_field, value, join_info.opts)
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/db/models/sql/query.py", line 1086, in check_related_objects
    for v in value:
  File "/home/chris/.virtualenvs/py3-dc-ynr/lib/python3.6/site-packages/django/utils/functional.py", line 257, in inner
    return func(self._wrapped, *args)
TypeError: 'AnonymousUser' object is not iterable
```

symroe replied:
I still can't reproduce this, but I've pulled the code out in to its own method now for easy of reading. This means the query will never get call for non-authenticated users anyway.

chris48s commented:
## High level stuff

I've left some detailed comments inline. Beyond that, this is going in a great direction. In particular, `BallotPaperView.get_context_data()` is much clearer and more readable. On the template side, this PR does a good job of bringing chunks of logic out of the templates and up into the model/view layers. On the downside, we've basically taken a big long sprawling HTML template and turned it into another slightly different big long sprawling HTML template (in fact, by merging `_candidates_for_post` into `ballot_view`, we've arguably made this slightly worse). I think there would be a lot of value in splitting this up into a number of smaller templates that we can pull in with `{% include %}`. For example, if we were to split out the blocks for "locking", "suggest lock", "previous candidates", "not standing" etc into fragments that we can include, that would be a huge win for readability here and very easy to do. I know we don't want to get too bogged down in HTML here, but I think separating these out would actually make the design work easier rather than harder.

---

## Misc stuff that doesn't fit in a line diff

This heading doesn't need to be here twice:

![Screenshot at 2019-08-21 16-17-09](https://user-images.githubusercontent.com/6025893/63519317-a750ea00-c4ea-11e9-8333-ee5c349b4634.png)

Also, this is showing me the suggest lock form when we have 0 candidates. Is locking with zero candidates ever valid?

---

Not sure what this is supposed to look like, but I'm guessing.. not this :D

![Screenshot at 2019-08-22 09-10-25](https://user-images.githubusercontent.com/6025893/63519334-ad46cb00-c4ea-11e9-9404-4fb2dccdef08.png)

---

We can ignore this as an issue for redesign if you want, but any reason Lock candidates is a "button small" here and the other are just "button". Maybe this is just a cue to think about when/where we use different button types/styles?

![Screenshot at 2019-08-21 16-30-15](https://user-images.githubusercontent.com/6025893/63519342-b20b7f00-c4ea-11e9-8737-ae333801b1f7.png)

---

The "Suggest locking" from is shown even if I'm not logged in - this should only be shown to authenticated users.

This isn't _strictly_ related, but worth thinking about in relation to this.. in general, we're using this pattern quite a lot:

```
# views.py
context["do_stuff_form"] = DoStuffForm()
context["user_can_do_stuff"] = user_can_do_stuff()

# template
{% if user_can_do_stuff %}
  {{ do_stuff_form }}
{% endif %}
```

I think it makes more sense to try and do it like this instead wherever possible:

```
# views.py
if user_can_do_stuff():
    context["do_stuff_form"] = DoStuffForm()

# template
{{ do_stuff_form }}
```

The second way is less total code and importantly keeps application logic out of the templates. I know that on this page we're quite often building up forms manually in the template, so we are going to have some `{% if user_can_do_stuff %}` checks in the template, but it feels like its worth trying to minimise that where we can. What do you think to that?

---

There are a bunch of things I should have tested in this PR, but I've struggled to reproduce all the scenarios I wanted to check. Unfortunately I've spent a lot of time today trying to use `candidates_import_from_live_site` to get a decent set of test data imported (spoiler: I failed), so there's a bunch of stuff I haven't been able to check.

symroe commented:
> Also, this is showing me the suggest lock form when we have 0 candidates. Is locking with zero candidates ever valid?

Uncontested seats do happen, and the chances are someone will spot this on YNR before we set the `uncontested` flag in EE/YNR. I think it's useful for someone to signal that they've seen an uncontested election – but maybe this needs a wording change or something?

> Not sure what this is supposed to look like, but I'm guessing.. not this :D

:D 93ec31aab9f

> We can ignore this as an issue for redesign if you want, but any reason Lock candidates is a "button small" here and the other are just "button". Maybe this is just a cue to think about when/where we use different button types/styles?

No real reason, no. I think the normal buttons are too large, but that's more of a design question. Can change if you like, but it's not a massive problem I reckon.

symroe commented:
> This isn't strictly related, but worth thinking about in relation to this.. in general, we're using this pattern quite a lot

Yes, I agree the second is better. I think there is more of that in this change, but some still remain.

The main one, I think, is `membership_edits_allowed`, however I think it's worth keeping this one, as quite a lot of the page changes depending on it's value. Showing / hiding the "log in to add new candidate" button, for example.

Are there others you think I should change?

chris48s commented:
> > Also, this is showing me the suggest lock form when we have 0 candidates. Is locking with zero candidates ever valid?
>
> Uncontested seats do happen

An uncontested seat would have one candidate though, not zero

symroe commented:
> An uncontested seat would have one candidate though, not zero

Well it would be either, right? But I agree that it's more useful to hide this when there aren't any candidates. I've done that in 3d0d95417313

chris48s commented on ynr/apps/elections/views.py:
```diff
@@ -138,830 +124,99 @@ class BallotPaperView(TemplateView):
-        context["redirect_after_login"] = context["ballot"].get_absolute_url()
+    def get_authenticated_user_context(self, ballot, context):
+        """
+        Stuff we only want to add if the user is authrnticated
```
```suggestion
        Stuff we only want to add if the user is authenticated
```

chris48s commented:
Cool. This is looking much better this time round. I think showing/hiding the entire `candidates_might_stand` / `candidates_not_standing` tables based on permissions is less error prone and fiddly than trying to selectively show/hide columns. Splitting `ballot_view` into smaller chunks makes it much more digestable IMO. :+1:

A few more comments, but we're pretty close to finished:

---

You've already acknowledged there is a failing test to sort out.

---

Just giving this a second read, in the view, we're setting:
`context["lock_form"] = ToggleLockForm(...`
but then we don't actually appear to be using `lock_form` in the template. `_ballot_lock_form.html` is building it up as custom HTML.
Can you have a look at that and either remove `context["lock_form"] = ToggleLockForm(...` if it is dead code, or replace `_ballot_lock_form.html` with `{{ lock_form }}` if its unnecessary duplication.

---

This "Suggest Locking" header is out-of-place in this context. Can we move it into the block with the form?

![Screenshot at 2019-09-04 11-26-45](https://user-images.githubusercontent.com/6025893/64249288-3eff0100-cf0b-11e9-9cd1-e802676d8d24.png)


---

Commits f13c07f, 9c1ad5f, bce5e6c and e563d3e are the commits from #1003, #1004 and #1005 but with different commit hashes. Did something go wrong with a rebase somewhere. I think this needs tidying up as it may cause problems merging this branch to master.

symroe commented:
* Toggle lock: Removed from context
* Tests: passing
* "Suggest Locking" header: fixed
* Commits: Hmm. I'm not sure TBH. I rebased this from origin/master after merging the Django 2.2 PR. Not sure what to do about this?

chris48s commented:
To take one commit as an example, what is the commit hash of "Better validation for death dates" on your local master branch? Is it 78866a4 or e563d3e ?

If I check this branch out and rebase it onto `upstream/master`, that seems to fix the history locally for me (my `upstream` remote is DemocracyClub/yournextrepresentative):

```
$ git fetch upstream

$ git log --oneline | grep "Better validation"
e563d3ea Better validation for death dates

$ git rebase upstream/master
First, rewinding head to replay your work on top of it...
<snip>

$ git log --oneline | grep "Better validation"
78866a41 Better validation for death dates
```

symroe commented:
On my local `master`, that commit is 78866a41

chris48s commented:
rebasing this branch onto master should do it then. See above code snippet

symroe commented:
There we go!

symroe commented:
Ace, thanks for all your time reviewing this!
